proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,727 @@
+
+lemma valid_cur_vcpu_lift_cur_thread_update:
+  assumes arch_tcb_at: "\<And>P. f \<lbrace>arch_tcb_at P t\<rbrace>"
+  and active_cur_vcpu_of: "\<And>P. f \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  shows "f \<lbrace>\<lambda>s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  apply (clarsimp simp: valid_cur_vcpu_def valid_def)
+  using use_valid[OF _ active_cur_vcpu_of] use_valid[OF _ arch_tcb_at]
+  by (fastforce simp: active_cur_vcpu_of_def)
+
+crunches as_user
+  for active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
+  (wp: crunch_wps simp: active_cur_vcpu_of_def)
+
+lemma as_user_valid_cur_vcpu[wp]:
+  "as_user tptr f \<lbrace>valid_cur_vcpu\<rbrace>"
+  apply (rule valid_cur_vcpu_lift)
+   apply (rule hoare_weaken_pre)
+   by (wpsimp wp: as_user_pred_tcb_at | wps)+
+
+lemma machine_state_update_active_cur_vcpu_of[simp]:
+  "P (active_cur_vcpu_of (s\<lparr>machine_state := ms\<rparr>)) = P (active_cur_vcpu_of s)"
+  by (fastforce simp: active_cur_vcpu_of_def)
+
+crunches do_machine_op
+  for valid_cur_vcpu[wp]: valid_cur_vcpu
+  and valid_cur_vcpu_cur_thread_update[wp]: "\<lambda>s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)"
+  (wp: valid_cur_vcpu_lift_cur_thread_update valid_cur_vcpu_lift crunch_wps)
+
+lemma valid_cur_vcpu_vcpu_update[simp]:
+  "vcpu_at v s \<Longrightarrow> valid_cur_vcpu (s\<lparr>kheap := kheap s(v \<mapsto> ArchObj (VCPU vcpu))\<rparr>) = valid_cur_vcpu s"
+  by (clarsimp simp: valid_cur_vcpu_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+
+crunches vcpu_save_reg, vcpu_write_reg, save_virt_timer, vgic_update, vcpu_disable
+  for valid_cur_vcpu[wp]: valid_cur_vcpu
+  (wp: set_vcpu_wp)
+
+lemma set_vcpu_arch_tcb_at_cur_thread[wp]:
+  "set_vcpu ptr vcpu \<lbrace>\<lambda>s. arch_tcb_at P (cur_thread s) s\<rbrace>"
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: pred_tcb_at_def obj_at_def)
+  done
+
+crunches vcpu_disable, vcpu_restore, vcpu_save, set_vm_root
+  for arch_tcb_at_cur_thread[wp]: "\<lambda>s. arch_tcb_at P (cur_thread s) s"
+  (wp: crunch_wps)
+
+crunches vcpu_update, do_machine_op, invalidate_hw_asid_entry, invalidate_asid
+  for active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
+  (simp: active_cur_vcpu_of_def)
+
+lemma vcpu_save_reg_active_cur_vcpu_of[wp]:
+  "vcpu_save_reg vr reg \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  by (wpsimp simp: vcpu_save_reg_def)
+
+crunches vcpu_restore, do_machine_op, vcpu_save_reg, vgic_update, save_virt_timer,
+         vcpu_save_reg_range, vgic_update_lr, vcpu_enable, vcpu_save
+  for valid_cur_vcpu[wp]: valid_cur_vcpu
+  (wp: valid_cur_vcpu_lift crunch_wps simp: active_cur_vcpu_of_def)
+
+lemma switch_vcpu_valid_cur_vcpu_cur_thread_update[wp]:
+  "\<lbrace>arch_tcb_at (\<lambda>itcb. itcb_vcpu itcb = v) t\<rbrace>
+   vcpu_switch v
+   \<lbrace>\<lambda>_ s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  apply (clarsimp simp: vcpu_switch_def)
```

Comment:
this one also looks like it could be an `unfolding` (it's not super important, but it's nice when the unfolding step is stable and you know in maintenance that nothing interesting is supposed to be happening here).

Reply:
This feels like a specific enough guideline that it would be worth recording. Would [`docs/style`](https://github.com/seL4/l4v/blob/master/docs/Style.thy#L251) be the right place to put it?

Reply:
Yes, that would be a good place